File: lib/common_words.rb
def most_common(string)
  words = string.split(' ')
  word_counts = Hash.new(0)
  punctuation = ['.', ',', '!', '?', ';', ':']
  words.each do |word|
    word.slice!(-1) if punctuation.any? { |punct| punct == word[-1] }
    word_counts[word.downcase] += 1
  end
  word_counts = word_counts.sort_by { |k, v| -v }
  # Calling 'sort_by' on a hash returns an array of arrays with key-value pairs.
  most_common_words = []
  word_counts.each do |word_and_frequency|
    if word_and_frequency[1] == word_counts.first[1]
      most_common_words << word_and_frequency[0]
    end
  end
  most_common_words
end

most_common('This is a string of words, it has punctuation and more than one of the word this this this.')
most_common("test test test greg jordan mark mark mark")

File: lib/common_words.js
var mostCommonWords = function(string){
  var words = string.split(' ');
  var word_counts = {};

  for(var i = 0; i < words.length; ++i){
    if(words[i][words[i].length - 1] === '.' || words[i][words[i].length - 1] === ','
      || words[i][words[i].length - 1] === '!' || words[i][words[i].length - 1] === '?'
      || words[i][words[i].length - 1] === '.' || words[i][words[i].length - 1] === ':'
      || words[i][words[i].length - 1] === ';'){
      words[i] = words[i].slice(0, -1);
    }
    word_counts[words[i].toLowerCase()] = 0;
  }

  for(i = 0; i < words.length; ++i){
    word_counts[words[i].toLowerCase()]++;
  }

  console.log(word_counts);
}

mostCommonWords('This is a string. It uses some words more often than others. Like this.')
mostCommonWords('There are three occurences of the word three in this string. Three times I say!  Say what? What is going on?')
